File: opensrp-chw-referral/src/main/java/org/smartregister/chw/referral/contract/BaseIssueReferralContract.java
package org.smartregister.chw.referral.contract;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import android.content.Context;

import org.apache.commons.lang3.tuple.Triple;
import org.jetbrains.annotations.Nullable;
import org.json.JSONObject;
import org.smartregister.chw.referral.domain.MemberObject;
import org.smartregister.chw.referral.domain.ReferralServiceIndicatorObject;
import org.smartregister.chw.referral.domain.ReferralServiceObject;
import org.smartregister.domain.Location;
import org.smartregister.view.contract.BaseRegisterContract;

import java.util.List;

public interface BaseIssueReferralContract {

    interface View {
        Presenter presenter();

        Context getCurrentContext();

        void setProfileViewWithData();
    }

    interface Presenter extends BaseRegisterContract.Presenter {
        <T extends ViewModel & Model> Class<T> getViewModel();

        boolean validateValues(MemberObject memberObject);

        String getMainCondition();

        String getMainTable();

        void fillClientData(@Nullable MemberObject memberObject);

        void initializeMemberObject(MemberObject memberObject);

        void saveForm(String jsonString);
    }

    interface Model {
        String getLocationId(String locationName);

        JSONObject getFormWithValuesAsJson(String formName, String entityId,
                                           String currentLocationId, MemberObject memberObject) throws Exception;

        String mainSelect(String tableName, String mainCondition);

        LiveData<List<Location>> getHealthFacilities();

        LiveData<List<ReferralServiceObject>> getReferralServicesList(List<String> referralServiceIds);

        List<ReferralServiceIndicatorObject> getIndicatorsByServiceId(String serviceId);

    }

    interface Interactor {

        void onDestroy(boolean isChangingConfiguration);

        void saveRegistration(String jsonString, final InteractorCallBack callBack);

    }

    interface InteractorCallBack {

        void onUniqueIdFetched(Triple<String, String, String> triple, String entityId);

        void onNoUniqueId();

        void onRegistrationSaved();

    }
}

File: opensrp-chw-referral/src/main/java/org/smartregister/chw/referral/contract/BaseIssueReferralContract.kt
package org.smartregister.chw.referral.contract

import androidx.lifecycle.ViewModel
import com.nerdstone.neatformcore.domain.model.NFormViewData
import org.apache.commons.lang3.tuple.Triple
import org.json.JSONObject
import org.koin.core.KoinComponent
import org.smartregister.chw.referral.domain.MemberObject
import org.smartregister.chw.referral.domain.ReferralServiceIndicatorObject
import org.smartregister.chw.referral.domain.ReferralServiceObject
import org.smartregister.domain.Location
import org.smartregister.view.contract.BaseRegisterContract
import java.util.*

interface BaseIssueReferralContract {

    interface View: KoinComponent {

        fun presenter(): Presenter

        fun setProfileViewWithData()

    }

    interface Presenter : BaseRegisterContract.Presenter {

        fun getView(): View?

        fun <T> getViewModel(): Class<T> where T : ViewModel, T : Model

        fun getMainCondition(): String

        fun getMainTable(): String

        fun fillClientData(memberObject: MemberObject?)

        fun initializeMemberObject(memberObject: MemberObject?)

        fun saveForm(valuesHashMap: HashMap<String, NFormViewData>, jsonObject: JSONObject)
    }

    interface Model {

        fun getLocationId(locationName: String?): String?

        fun mainSelect(tableName: String, mainCondition: String): String

        val healthFacilities: List<Location>?

        fun getReferralServicesList(referralServiceId: String): ReferralServiceObject?

        fun getIndicatorsByServiceId(serviceId: String): List<ReferralServiceIndicatorObject>?
    }

    interface Interactor: KoinComponent {

        fun onDestroy(isChangingConfiguration: Boolean)

        fun saveRegistration(
            baseEntityId: String, valuesHashMap: HashMap<String, NFormViewData>,
            jsonObject: JSONObject, callBack: InteractorCallBack
        )
    }

    interface InteractorCallBack {

        fun onUniqueIdFetched(triple: Triple<String, String, String>, entityId: String)

        fun onNoUniqueId()

        fun onRegistrationSaved()
    }
}